### ISSUE TITLE
Host simulation hooks with extended abilities

### ISSUE THREAD

**Issue by jiruhameru** [FLAMEGPU/FLAMEGPU#177]
Seems there is an incompatibility between host and device code:
---------------------------------------------------------------------------------------------
errors: 

> calling a __host__ function("updateLogFile") from a __device__ function("agent_output_location") is not allowed
> 
> error: calling a __host__ function("fopen") from a __device__ function("____") is not allowed
> 
> error: identifier "fopen" is undefined in device code
> 
> 

---------------------------------------------------------------------------------------------
I've been trying to fetch all agents' data and write them onto an external file. Unfortunately, the only currently available analytic functions are: reduce and count functions & they are insufficient.

I need all agents' data during the whole simulation for a more specialized data-analysis, so it'd be very
helpful to add this functionality, or enlighten us with a decent solution.

Thanks in advance.

**Comment by ptheywood** [comment]
Within CUDA applications such as FLAME GPU, functions can be defined as `__host__` or `__device__`.
Host functions execute on the CPU, whilst Device functions execute on the GPU. 
If neither is present, the function is implicitly a `__host__` function.

+ `__host__` functions can only call other `__host__` functions (and `__global__` kernels)
+ `__device__` functions can only contain calls to other `__device__` functions. 

Within the context of FLAME GPU, agent functions are labelled with the macro `__FLAME_GPU_FUNC__`, this maps to `__device__`, and so these functions execute on the GPU and cannot include any calls to `__host__` functions such as `updateLogFile`.


When running console mode FLAME GPU models, by default an XML file is produced at the end of each simulation iteration. The XML file is output to the same directory as the initial states file, and is name `<iteration_number>.xml`, i.e. `1.xml`, `2.xml`, ...
You could then write a separate application which reads one or more XML files from disk, and carries out your more specialised data-analysis.

For the next version of FLAME GPU (1.5.0), we have added some functionality for creating agents on the host and transferring them to the device, however we've not yet added the ability to access agent data from host functions. 
This does seem like a useful feature, so I will create a new issue so it is simple for users to write more complex analytics in host functions.

**Comment by ptheywood** [comment]
Having re-read your issue

>I've been trying to fetch all agents' data and write them onto an external file. Unfortunately, the only currently available analytic functions are: reduce and count functions & they are insufficient.
>
> I need all agents' data during the whole simulation for a more specialized data-analysis, so it'd be very
helpful to add this functionality, or enlighten us with a decent solution.

Currently if you use the console mode, all agent data will be output to the host at each iteration, as `iterations/<iteration>.xml` (if using `iterations/0.xml` as the input file). You can then use an offline process to load values from each xml file and perform your analysis. This could be using R, python or even spreadsheets.

Alternatively, when #192 is merged into `release-1.5`, and subsequently `master` when version `1.5.0` is released it will be possible to write custom `STEP` functions which can access agent variables on the host.

For instance, with an agent `foo` in the state `bar`, and variables:

```
int id
float x
float y
float value_a
float value_b
float value_c
float value_d
``` 


It would be possible to write a function to output a CSV file per simulation iteration containing only the `id`, `x` and `y` variables as follows: 


```
__FLAME_GPU_STEP_FUNC__ void outputCSV(){

    // Get some values and construct an output path.
    const char * directory = getOutputDir();
    unsigned int iteration = getIterationNumber();

    std::string outputFilename = std::string( std::string(directory) + "iteration-" + std::to_string(iteration) +".csv");

    // Get a file handle for output.
    FILE * fp = fopen(outputFilename.c_str(), "w");
    // If the file has been opened successfully
    if(fp != nullptr){
        fprintf(stdout, "Outputting csv to %s\n", outputFilename.c_str());

        // Output a header row for the CSV
        fprintf(fp, "id, x, y\n");

        // For each foo in state bar
        for(int index = 0; index < get_agent_foo_bar_count(); index++){
            // Append a row to the CSV file
            fprintf(
                fp, 
                "%u, %f, %f\n",
                get_foo_bar_variable_id(index),
                get_foo_bar_variable_x(index),
                get_foo_bar_variable_y(index)
            );
        }
        // Flush the file handle
        fflush(fp);
    } else {
        // Issue an error message
        fprintf(stderr, "Error: file %s could not be created for outputCSV\n", outputFilename.c_str());
    }
    // Close the file handle if necessary.
    if (fp != nullptr){
        fclose(fp);
        fp = nullptr;
    }
}
```

It could output all of the agent properties, however that can already be achieved by post-processing the generated XML files. 

Agent variable arrays are also supported, with an extra parameter to the `get_*_variables_*` function of `element`, indicating which `element` of the array should be accessed.

i.e to access the `4`th element of the `arr`  variable for the `512`th agent of type `foo` in state `bar`:

```
unsigned int index = 512;
unsigned int element = 4;

float value = get_foo_bar_variable_arr(index, element);
```

**However**, care **must** be taken when using these functions. These functions implicitly copy data from the `device` to the `host` over the PCI-e bus. This is an **expensive** process, so this should only be performed when necessary.



Feel free to re-open this issue if you have further questions.
